[CL-59] [Bug] Link buttons have different height depending on html tag used

coroiu opened the pull request:
## Type of change

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

`bitLink` using the `button` dom element do not line up properly with surrounding text.

## Code changes
 
Buttons are automatically set to `inline-block` (and cannot be changed to `inline`) causing height calculations to happen differently. Setting an explicit line height to match surrounding text fixes this.

## Screenshots

### Without fix
![image](https://user-images.githubusercontent.com/2285588/199695861-05b81edc-a5fb-4a82-98f7-493308f3134b.png)

### With fix
![image](https://user-images.githubusercontent.com/2285588/199695958-b9f190c8-1364-4520-a3ea-2e02dc9267a1.png)

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

danielleflinn commented:
It looks like the link-button lost some padding and the link-anchor still includes it. Is that intentional? 

Padding in PR
<img width="200" alt="Screen Shot 2022-11-03 at 8 22 54 AM" src="https://user-images.githubusercontent.com/43477473/199762614-d51275e1-6102-4312-abc3-da71b4553d88.png">

Padding in components.bitwarden.com
<img width="191" alt="Screen Shot 2022-11-03 at 8 25 20 AM" src="https://user-images.githubusercontent.com/43477473/199762756-53333c2a-5af0-472a-a231-3b590caebc69.png">

vincentsalucci commented:
I apologize - I didn't see @danielleflinn 's comments above. Make sure they're addressed before merging 👍

coroiu commented:
Nope that wasn't intentional, I'll get back to this when I have some time in-between tasks

coroiu commented:
@danielleflinn I've taken a closer look at this and the fix was not as simple as I initially thought. I think I have something working which you can simply review in storybook. For a more detailed explanation you can keep reading :) 

The main problem is apparently that buttons cannot be made into `display: inline`, browsers just ignore that rule (see https://github.com/w3c/csswg-drafts/issues/3226). The best you can get is `display: inline-block` (the default) which of course differs in some key aspects, mainly in the way height and padding is calculated. The consequence is that we cannot use simple padding to push away the outline on the button (and if we try we push the whole button downwards). The two html elements also currently differ vastly in how they treat `line-height` on the parent element.

The best workaround I’ve found is to split up the directives and use different methods of outlining. On the button we can instead use the same “hack” that is currently used on the icon buttons. There will be no change to the external interface, meaning `<a bitLink` and `<button bitLink` will both continue to work like before. By doing this we get all `bitLink` elements aligned on the baseline meaning that we can combine them freely in the same sentence/row (which is done inside the breadcrumbs). 

The downside is increased complexity and that the outline offset on the two elements seems to diverge when text size gets bigger. Though this is probably never gonna be noticeable since there will never be two outlines both active and in close enough proximity to each other to be noticeable.

## Screenshots (with activated `focus-visible` styles for easier comparison)

### Current solution 
![image](https://user-images.githubusercontent.com/2285588/201337541-2475c0fe-8398-432b-95fe-855cefab468e.png)

### With workaround
![image](https://user-images.githubusercontent.com/2285588/201337698-cd197962-c8fe-46f8-b535-737284139f52.png)

danielleflinn commented:
> By doing this we get all bitLink elements aligned on the baseline meaning that we can combine them freely in the same sentence/row (which is done inside the breadcrumbs)...The downside is increased complexity

The fix seems to behave as we were wanting. That said, @coroiu are you concerned about the additional complexity the fix adds?

coroiu commented:
@danielleflinn When complexity is extremely isolated to a single component like this I tend to have a pretty high tolerance for complexity - a limit which I don't think this solution is close to reaching. But I'd like to get some feedback on the code from another developer as well, maybe @Hinton ? :D

Hinton reviewed (commented):
Do we have any story showcasing the link inline?

And have we tested this in different browsers?

coroiu commented:
I added some extra padding that I seem to have missed and which was putting the outline very tightly around the text

> Do we have any story showcasing the link inline?

Good point, I've added a story containing an example.

> And have we tested this in different browsers?

Yes, here are some screenshots

### Firefox
![image](https://user-images.githubusercontent.com/2285588/202472559-4214bd8a-ef58-4375-a49e-e1a6bb25b768.png)

### Chrome
![image](https://user-images.githubusercontent.com/2285588/202472736-b4b88762-e8c2-4399-97e8-9ffa4544a11c.png)

### Safari
![image](https://user-images.githubusercontent.com/2285588/202472919-1d74b6bd-c3f6-47b4-a3a4-72c422522051.png)

### Internet explorer
Nope 😄
